Patch-Source: https://github.com/brailcom/speechd/commit/32499092928d5bbefa4815eb0fa082e840ba5c7f
--
From 32499092928d5bbefa4815eb0fa082e840ba5c7f Mon Sep 17 00:00:00 2001
From: David Turner <novalis@novalis.org>
Date: Thu, 1 Jun 2023 20:36:43 -0400
Subject: [PATCH] Speed up spd_execute_command_with_list_reply

spd_execute_command_with_list_reply was accidentally quadratic: it
would start by searching the string for one \r\n, then it would search
for two, then for three...  I have 15,000 voices (I have no idea why;
it seems to be the default Debian config), so this was quite slow.

For me, this improves the performance of spd_list_synthesis_voices
from 1.5s to 0.5s.

Signed-off-by: David Turner <novalis@novalis.org>
---
 src/api/c/libspeechd.c | 49 +++++++++++++++++++++++++++---------------
 1 file changed, 32 insertions(+), 17 deletions(-)

diff --git a/src/api/c/libspeechd.c b/src/api/c/libspeechd.c
index 6d2aabfd..1b74889e 100644
--- a/src/api/c/libspeechd.c
+++ b/src/api/c/libspeechd.c
@@ -70,6 +70,7 @@ static int isanum(char *str);
 static char *get_reply(SPDConnection * connection);
 static int get_err_code(char *reply);
 static char *get_param_str(char *reply, int num, int *err);
+static char *get_param_str_and_advance(char **reply, int *err);
 static int get_param_int(char *reply, int num, int *err);
 static int ret_ok(char *reply);
 static void SPD_DBG(char *format, ...);
@@ -1542,8 +1543,9 @@ char **spd_execute_command_with_list_reply(SPDConnection * connection,
 
 	result = malloc((max_items + 1) * sizeof(char *));
 
+	char *cur = reply;
 	for (i = 0;; i++) {
-		line = get_param_str(reply, i + 1, &err);
+		line = get_param_str_and_advance(&cur, &err);
 		if ((err) || (line == NULL))
 			break;
 		result[i] = line;
@@ -1986,9 +1988,8 @@ static int ret_ok(char *reply)
 	SPD_FATAL("Internal error during communication.");
 }
 
-static char *get_param_str(char *reply, int num, int *err)
+static char *get_param_str_and_advance(char **reply, int *err)
 {
-	int i;
 	char *tptr;
 	char *pos;
 	char *pos_begin;
@@ -1997,20 +1998,7 @@ static char *get_param_str(char *reply, int num, int *err)
 
 	assert(err != NULL);
 
-	if (num < 1) {
-		*err = -1;
-		return NULL;
-	}
-
-	pos = reply;
-	for (i = 0; i <= num - 2; i++) {
-		pos = strstr(pos, "\r\n");
-		if (pos == NULL) {
-			*err = -2;
-			return NULL;
-		}
-		pos += 2;
-	}
+	pos = *reply;
 
 	if (strlen(pos) < 4)
 		return NULL;
@@ -2034,9 +2022,36 @@ static char *get_param_str(char *reply, int num, int *err)
 	rep = (char *)strndup(pos_begin, pos_end - pos_begin);
 	*err = 0;
 
+	*reply = pos_end + 2;
+
 	return rep;
 }
 
+static char *get_param_str(char *reply, int num, int *err)
+{
+	int i;
+	char *pos;
+
+	assert(err != NULL);
+
+	if (num < 1) {
+		*err = -1;
+		return NULL;
+	}
+
+	pos = reply;
+	for (i = 0; i <= num - 2; i++) {
+		pos = strstr(pos, "\r\n");
+		if (pos == NULL) {
+			*err = -2;
+			return NULL;
+		}
+		pos += 2;
+	}
+
+	return get_param_str_and_advance(&pos, err);
+}
+
 static int get_param_int(char *reply, int num, int *err)
 {
 	char *rep_str;
